persistence/docs/PROTOCOL_STATE_HASH.md

```diff
@@ -0,0 +1,131 @@
+
+### Block Proto
+
+The block protobuf that is serialized and store in the block store can be found in `persistence/proto/block_persistence.proto`. This proto contains the `stateHash` along with the corresponding height.
+
+### Trees
+
+An individual Merkle Tree is created for each type of actor, record or data type. Each of these is backed by its own key-value store.
+
+**Actor Merkle Trees**:
+
+- Applications
+- Validators
+- Fisherman
+- ServiceNodes
+
+**Account Merkle Trees**:
+
+- Accounts
+- Pools
+
+**Data Merkle Trees**
+
+- Transactions
+- Parameters
+- Flags
+
+## Compute State Hash
+
+_Note: `GetRecordsUpdatedAtHeight` is an abstraction for retrieving all the records from the corresponding SQL tables depending on the type of record (Actors, Transactions, Params, etc...)_
+
+This flow shows the interaction between the `PostgresDB` and `MerkleTrees` listed above to compute the state hash. Assuming the process of applying a proposal block to the current context (i.e. the uncommited SQL state) is done, the following steps compute the hash of the new world state.
+
+1. Loop over all of the merkle tree types
+2. Use `GetRecordsUpdatedAtHeight` to retrieve all the records updated at the context's height
+3. Serialize each record using the corresponding underlying protobuf
+4. Insert the serialized record into the corresponding tree (which is back by a key-value store)
+5. Compute the root hash of each tree
+6. Aggregate all the root hashes by concatenating them together
+7. Compute the new `stateHash` by taking a hash of the concatenated hash list
+
+```mermaid
+sequenceDiagram
+    participant P as Persistence
+    participant PSQL as Persistence (SQL Store)
+    participant PKV as Persistence (Key-Value Store)
+
+    loop for each merkle tree type
+        P->>+PSQL: GetRecordsUpdatedAtHeight(height, recordType)
+        PSQL->>-P: records
+        loop for each state tree
+            P->>+PKV: Update(addr, serialize(record))
+            PKV->>-P: result, err_code
+        end
+        P->>+PKV: GetRoot()
+        PKV->>-P: rootHash
+    end
+
+    P->>P: stateHash = hash(concat(rootHashes))
+    activate P
+    deactivate P
+```
+
+_IMPORTANT: The order in which the `rootHashes` are concatenated is based on the definition in which the trees are ordered in within `state.go`._
+
+## Store Block (Commit)
+
+When the `Commit(quorumCert)` function is invoked, the current context is committed to disk. The `PersistenceContext` does the following:
+
+1. Read data from the persistence context's in-memory state
+2. Prepare a instance of the `Block` proto & serialize it
+3. Insert the `Block` into the `BlockStore`
+4. Insert the `Block` into the SQL Store
+5. Commit the context's SQL transaction to disk
```

Comment:
What about the unhappy path / rollback? Out of scope for now? Do we have issues/TODOs for it?

Reply:
LOL nevermind maybe answered myself with the below

Reply:
Added this section
```
## Failed Commitments

TODO: Failed commitments and the implementation of rollbacks is tracked in #327 and #329.
```